File: dataset/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Task131.java

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import java.security.*;

public class Task131 {
    private static final String MAGIC = "ENC1";
    private static final int VERSION = 1;
    private static final int SALT_SIZE = 16;
    private static final int IV_SIZE = 12;
    private static final int TAG_SIZE = 16;
    private static final int MAX_DATA_SIZE = 1024 * 1024; // 1MB limit
    
    public static class SecureData implements Serializable {
        private static final long serialVersionUID = 1L;
        private final String data;
        private final long timestamp;
        
        public SecureData(String data) {
            if (data == null || data.length() > 10000) {
                throw new IllegalArgumentException("Invalid data");
            }
            this.data = data;
            this.timestamp = System.currentTimeMillis();
        }
        
        public String getData() { return data; }
        public long getTimestamp() { return timestamp; }
        
        @Override
        public String toString() {
            return "SecureData{data='" + data + "', timestamp=" + timestamp + "}";
        }
    }
    
    public static byte[] serializeData(SecureData data, String passphrase) throws Exception {
        if (data == null || passphrase == null || passphrase.length() < 8) {
            throw new IllegalArgumentException("Invalid input");
        }
        
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        baos.write(MAGIC.getBytes(StandardCharsets.UTF_8));
        baos.write(VERSION);
        
        byte[] salt = new byte[SALT_SIZE];
        SecureRandom random = new SecureRandom();
        random.nextBytes(salt);
        baos.write(salt);
        
        SecretKeySpec key = deriveKey(passphrase, salt);
        byte[] iv = new byte[IV_SIZE];
        random.nextBytes(iv);
        baos.write(iv);
        
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec spec = new GCMParameterSpec(TAG_SIZE * 8, iv);
        cipher.init(Cipher.ENCRYPT_MODE, key, spec);
        
        ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
        try (ObjectOutputStream oos = new ObjectOutputStream(dataStream)) {
            oos.writeObject(data);
        }
        byte[] plaintext = dataStream.toByteArray();
        
        byte[] ciphertext = cipher.doFinal(plaintext);
        baos.write(ciphertext);
        
        return baos.toByteArray();
    }
    
    public static SecureData deserializeData(byte[] encryptedData, String passphrase) throws Exception {
        if (encryptedData == null || passphrase == null || passphrase.length() < 8) {
            throw new IllegalArgumentException("Invalid input");
        }
        
        if (encryptedData.length > MAX_DATA_SIZE) {
            throw new IllegalArgumentException("Data too large");
        }
        
        ByteArrayInputStream bais = new ByteArrayInputStream(encryptedData);
        
        byte[] magic = new byte[4];
        if (bais.read(magic) != 4 || !Arrays.equals(magic, MAGIC.getBytes(StandardCharsets.UTF_8))) {
            throw new SecurityException("Invalid magic");
        }
        
        int version = bais.read();
        if (version != VERSION) {
            throw new SecurityException("Invalid version");
        }
        
        byte[] salt = new byte[SALT_SIZE];
        if (bais.read(salt) != SALT_SIZE) {
            throw new SecurityException("Invalid salt");
        }
        
        byte[] iv = new byte[IV_SIZE];
        if (bais.read(iv) != IV_SIZE) {
            throw new SecurityException("Invalid IV");
        }
        
        byte[] ciphertext = bais.readAllBytes();
        if (ciphertext.length < TAG_SIZE) {
            throw new SecurityException("Invalid ciphertext");
        }
        
        SecretKeySpec key = deriveKey(passphrase, salt);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec spec = new GCMParameterSpec(TAG_SIZE * 8, iv);
        cipher.init(Cipher.DECRYPT_MODE, key, spec);
        
        byte[] plaintext = cipher.doFinal(ciphertext);
        
        try (ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(plaintext)) {
            @Override
            protected Class<?> resolveClass(ObjectStreamClass desc) throws IOException, ClassNotFoundException {
                if (!desc.getName().equals(SecureData.class.getName())) {
                    throw new SecurityException("Unauthorized deserialization attempt");
                }
                return super.resolveClass(desc);
            }
        }) {
            Object obj = ois.readObject();
            if (!(obj instanceof SecureData)) {
                throw new SecurityException("Invalid object type");
            }
            return (SecureData) obj;
        }
    }
    
    private static SecretKeySpec deriveKey(String passphrase, byte[] salt) throws Exception {
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        KeySpec spec = new PBEKeySpec(passphrase.toCharArray(), salt, 100000, 256);
        SecretKey tmp = factory.generateSecret(spec);
        return new SecretKeySpec(tmp.getEncoded(), "AES");
    }
    
    public static void main(String[] args) {
        try {
            System.out.println("Test Case 1: Basic serialization/deserialization");
            SecureData data1 = new SecureData("Hello World");
            byte[] encrypted1 = serializeData(data1, "strongpassphrase123");
            SecureData decrypted1 = deserializeData(encrypted1, "strongpassphrase123");
            System.out.println("Original: " + data1);
            System.out.println("Decrypted: " + decrypted1);
            System.out.println("Match: " + data1.getData().equals(decrypted1.getData()));
            System.out.println();
            
            System.out.println("Test Case 2: Different data");
            SecureData data2 = new SecureData("Sensitive information 12345");
            byte[] encrypted2 = serializeData(data2, "anotherpassword456");
            SecureData decrypted2 = deserializeData(encrypted2, "anotherpassword456");
            System.out.println("Original: " + data2);
            System.out.println("Decrypted: " + decrypted2);
            System.out.println("Match: " + data2.getData().equals(decrypted2.getData()));
            System.out.println();
            
            System.out.println("Test Case 3: Wrong passphrase");
            try {
                SecureData data3 = new SecureData("Test data");
                byte[] encrypted3 = serializeData(data3, "correctpass123");
                deserializeData(encrypted3, "wrongpass123");
                System.out.println("ERROR: Should have thrown exception");
            } catch (Exception e) {
                System.out.println("Correctly rejected wrong passphrase: " + e.getClass().getSimpleName());
            }
            System.out.println();
            
            System.out.println("Test Case 4: Corrupted data");
            try {
                SecureData data4 = new SecureData("Test data");
                byte[] encrypted4 = serializeData(data4, "mypassword789");
                encrypted4[encrypted4.length - 1] ^= 0xFF;
                deserializeData(encrypted4, "mypassword789");
                System.out.println("ERROR: Should have thrown exception");
            } catch (Exception e) {
                System.out.println("Correctly rejected corrupted data: " + e.getClass().getSimpleName());
            }
            System.out.println();
            
            System.out.println("Test Case 5: Empty string");
            SecureData data5 = new SecureData("");
            byte[] encrypted5 = serializeData(data5, "emptytest123");
            SecureData decrypted5 = deserializeData(encrypted5, "emptytest123");
            System.out.println("Original: " + data5);
            System.out.println("Decrypted: " + decrypted5);
            System.out.println("Match: " + data5.getData().equals(decrypted5.getData()));
            
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: dataset/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Task131.py

import struct
import os
import hashlib
import pickle
from typing import Any
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes
from dataclasses import dataclass
import time

MAGIC = b"ENC1"
VERSION = 1
SALT_SIZE = 16
IV_SIZE = 12
TAG_SIZE = 16
MAX_DATA_SIZE = 1024 * 1024  # 1MB limit


@dataclass
class SecureData:
    data: str
    timestamp: float
    
    def __post_init__(self) -> None:
        if not isinstance(self.data, str) or len(self.data) > 10000:
            raise ValueError("Invalid data")


class RestrictedUnpickler(pickle.Unpickler):
    def find_class(self, module: str, name: str) -> type:
        if module == "__main__" and name == "SecureData":
            return SecureData
        raise pickle.UnpicklingError(f"Unauthorized class: {module}.{name}")


def derive_key(passphrase: str, salt: bytes) -> bytes:
    if not isinstance(passphrase, str) or len(passphrase) < 8:
        raise ValueError("Invalid passphrase")
    if not isinstance(salt, bytes) or len(salt) != SALT_SIZE:
        raise ValueError("Invalid salt")
    
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
    )
    return kdf.derive(passphrase.encode('utf-8'))


def serialize_data(data: SecureData, passphrase: str) -> bytes:
    if not isinstance(data, SecureData) or not isinstance(passphrase, str):
        raise ValueError("Invalid input")
    if len(passphrase) < 8:
        raise ValueError("Passphrase too short")
    
    output = bytearray()
    output.extend(MAGIC)
    output.append(VERSION)
    
    salt = os.urandom(SALT_SIZE)
    output.extend(salt)
    
    key = derive_key(passphrase, salt)
    iv = os.urandom(IV_SIZE)
    output.extend(iv)
    
    plaintext = pickle.dumps(data)
    
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, plaintext, None)
    output.extend(ciphertext)
    
    return bytes(output)


def deserialize_data(encrypted_data: bytes, passphrase: str) -> SecureData:
    if not isinstance(encrypted_data, bytes) or not isinstance(passphrase, str):
        raise ValueError("Invalid input")
    if len(passphrase) < 8:
        raise ValueError("Passphrase too short")
    if len(encrypted_data) > MAX_DATA_SIZE:
        raise ValueError("Data too large")
    
    min_size = len(MAGIC) + 1 + SALT_SIZE + IV_SIZE + TAG_SIZE
    if len(encrypted_data) < min_size:
        raise ValueError("Data too small")
    
    offset = 0
    
    magic = encrypted_data[offset:offset + len(MAGIC)]
    if magic != MAGIC:
        raise ValueError("Invalid magic")
    offset += len(MAGIC)
    
    version = encrypted_data[offset]
    if version != VERSION:
        raise ValueError("Invalid version")
    offset += 1
    
    salt = encrypted_data[offset:offset + SALT_SIZE]
    offset += SALT_SIZE
    
    iv = encrypted_data[offset:offset + IV_SIZE]
    offset += IV_SIZE
    
    ciphertext = encrypted_data[offset:]
    
    key = derive_key(passphrase, salt)
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext, None)
    
    import io
    obj = RestrictedUnpickler(io.BytesIO(plaintext)).load()
    
    if not isinstance(obj, SecureData):
        raise ValueError("Invalid object type")
    
    return obj


def main() -> None:
    print("Test Case 1: Basic serialization/deserialization")
    data1 = SecureData("Hello World", time.time())
    encrypted1 = serialize_data(data1, "strongpassphrase123")
    decrypted1 = deserialize_data(encrypted1, "strongpassphrase123")
    print(f"Original: {data1}")
    print(f"Decrypted: {decrypted1}")
    print(f"Match: {data1.data == decrypted1.data}")
    print()
    
    print("Test Case 2: Different data")
    data2 = SecureData("Sensitive information 12345", time.time())
    encrypted2 = serialize_data(data2, "anotherpassword456")
    decrypted2 = deserialize_data(encrypted2, "anotherpassword456")
    print(f"Original: {data2}")
    print(f"Decrypted: {decrypted2}")
    print(f"Match: {data2.data == decrypted2.data}")
    print()
    
    print("Test Case 3: Wrong passphrase")
    try:
        data3 = SecureData("Test data", time.time())
        encrypted3 = serialize_data(data3, "correctpass123")
        deserialize_data(encrypted3, "wrongpass123")
        print("ERROR: Should have thrown exception")
    except Exception as e:
        print(f"Correctly rejected wrong passphrase: {type(e).__name__}")
    print()
    
    print("Test Case 4: Corrupted data")
    try:
        data4 = SecureData("Test data", time.time())
        encrypted4 = bytearray(serialize_data(data4, "mypassword789"))
        encrypted4[-1] ^= 0xFF
        deserialize_data(bytes(encrypted4), "mypassword789")
        print("ERROR: Should have thrown exception")
    except Exception as e:
        print(f"Correctly rejected corrupted data: {type(e).__name__}")
    print()
    
    print("Test Case 5: Empty string")
    data5 = SecureData("", time.time())
    encrypted5 = serialize_data(data5, "emptytest123")
    decrypted5 = deserialize_data(encrypted5, "emptytest123")
    print(f"Original: {data5}")
    print(f"Decrypted: {decrypted5}")
    print(f"Match: {data5.data == decrypted5.data}")


if __name__ == "__main__":
    main()
